supabase/migrations/04_auth.sql

```diff
@@ -123,74 +125,88 @@ comment on column role_grants.capability is
+      where role_grants.subject_role ^@ all_roles.role_prefix
+        and role_grants.capability >= min_capability
         and all_roles.capability = 'admin'
   )
-  select role_prefix, capability from all_roles;
+  select role_prefix, max(capability) from all_roles
+  group by role_prefix
+  order by role_prefix;
 
-$$ language sql stable security definer;
+$$ language sql stable;
 
 
-create function auth_roles()
+create function auth_roles(min_capability grant_capability default 'x_00')
 returns table (role_prefix catalog_prefix, capability grant_capability) as $$
-  select role_prefix, capability from internal.user_roles(auth_uid())
+  select role_prefix, capability from internal.user_roles(auth_uid(), min_capability)
 $$ language sql stable security definer;
 comment on function auth_roles is
   'auth_roles returns all roles and associated capabilities of the user';
 
 
-create function auth_catalog(name_or_prefix text, min_cap grant_capability)
-returns boolean as $$
-  select exists(
-    select 1 from auth_roles() r
-    where starts_with(name_or_prefix, r.role_prefix) and r.capability >= min_cap
-  )
-$$ language sql stable;
-comment on function auth_catalog is
-  'auth_catalog returns true if the user has at least `min_cap` capability to the desired catalog `name_or_prefix`';
-
-
 -- Policy permissions for user_grants.
 create policy "Users select user grants they admin or are the subject"
   on user_grants as permissive for select
-  using (auth_catalog(object_role, 'admin') or user_id = auth.uid());
+  using (exists(
+    select 1 from auth_roles('admin') r where object_role ^@ r.role_prefix
+  ) or user_id = auth.uid());
 create policy "Users insert user grants they admin"
   on user_grants as permissive for insert
-  with check (auth_catalog(object_role, 'admin'));
+  with check (exists(
+    select 1 from auth_roles('admin') r where object_role ^@ r.role_prefix
+  ));
 create policy "Users update user grants they admin"
   on user_grants as permissive for update
-  using (auth_catalog(object_role, 'admin'));
+  using (exists(
+    select 1 from auth_roles('admin') r where object_role ^@ r.role_prefix
+  ));
 create policy "Users delete user grants they admin or are the subject"
   on user_grants as permissive for delete
-  using (auth_catalog(object_role, 'admin') or user_id = auth.uid());
+  using (exists(
+    select 1 from auth_roles('admin') r where object_role ^@ r.role_prefix
+  ) or user_id = auth.uid());
 
 grant all on user_grants to authenticated;
 
 
 -- Policy permissions for role_grants.
-create policy "Users select role grants they recieve or admin the object"
+create policy "Users select role grants where they admin the subject or object"
   on role_grants as permissive for select
-  using (auth_catalog(object_role, 'admin') or auth_catalog(subject_role, 'x_00'));
+  using (exists(
+    select 1 from auth_roles('admin') r
+    where (object_role ^@ r.role_prefix or subject_role ^@ r.role_prefix)
+  ));
```

Comment:
does inlining `catalog` help performance as well?

Reply:
I think you're referring to `auth_catalog()` calls here -- and the answer is a resounding "yes".

`auth_catalog(subject_role, 'admin')` is a SQL function, and I naively hoped / assumed that the optimizer would inline it, would see the call to auth_roles() which has no dependent arguments, and that auth_roles() is stable, and would then make a single call to auth_roles() for the query which is then used for evaluating every selected row examined by the policy.

What _actually_ happens, though, is that `auth_catalog(subject_role, 'admin')` is _not_ inlined. I'm pretty sure this is [because of its use of a sub-select](https://wiki.postgresql.org/wiki/Inlining_of_SQL_functions). Whatever the reason, it's not inlined and -- because it has an argument which is dependent on the current row -- it's re-evaluated for every single examined row of the selected table.

What happens now, is that a single call to auth_roles() is made for the lifetime of the query and its results are used over and over again with each examined row. That _does_ mean it's checking each row against the full set of authorized rows ever time -- ideally the optimizer would be able to invert this and identify subsets of the table to even look at in the first place -- but it's much much faster than re-computing auth_roles() every time.

The `combined_specs_ext` and `live_specs_ext` views are faster still, because they _do_ get inverted and the optimizer only bothers to look at rows of the base table which match your authorization scopes.

The included benchmark and `explain analyze` helped a lot in figuring all of this out.

Reply:
Just to expand, what I found is that the former `live_specs_ext` was certainly faster than a naive scan, because it avoided re-computing authorized roles with each role. _But_ it was still scanning every row of `live_specs` and comparing each against all of the user's roles.

Now it's faster, because the user's roles are used to determine which chunks of live_specs to look at.

Reply:
damn, I also assumed the function would get inlined. thanks for the explanation!